From 938ca0a20f12edd3474afab56b8fb3dac7289c5b Mon Sep 17 00:00:00 2001
From: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
Date: Wed, 24 May 2023 11:58:01 +0530
Subject: [PATCH] mac80211: WAR Placing station check before WARNON to avoid
 trace

Currently, power save is a link specific feature and from
ieee80211_recalc_ps_vif(), we are using ieee80211_vif_cfg_change_notify()
to notify the driver.

As power save is a link specific feature now, we remove power save flag
from vif specifc flags and hence it will trigger the WARNON from
ieee80211_vif_cfg_change_notify()

The trace is seen only when WDS is enabled, so keeping the station check
before the WARNON to avoid the trace.

Signed-off-by: Rahul Bhattacharjee <quic_rbhattac@quicinc.com>
---
 net/mac80211/main.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/net/mac80211/main.c
+++ b/net/mac80211/main.c
@@ -264,7 +264,8 @@ void ieee80211_vif_cfg_change_notify(str
 {
 	struct ieee80211_local *local = sdata->local;
 
-	WARN_ON_ONCE(changed & ~BSS_CHANGED_VIF_CFG_FLAGS);
+	if (sdata->u.mgd.use_4addr && sdata->vif.type != NL80211_IFTYPE_STATION)
+		WARN_ON_ONCE(changed & ~BSS_CHANGED_VIF_CFG_FLAGS);
 
 	if (!changed || sdata->vif.type == NL80211_IFTYPE_AP_VLAN)
 		return;
